lib/micromachine.rb

```diff
@@ -44,9 +49,21 @@ def states
 private
 
   def change(event)
-    @state = transitions_for[event][@state]
+    previous_state, @state = @state, transitions_for[event][@state]
+
+    all_arguments = [event, previous_state, state]
+
     callbacks = @callbacks[@state] + @callbacks[:any]
-    callbacks.each { |callback| callback.call(event) }
+    callbacks.each do |callback|
+      case callback.arity
+      when -1
+        callback.call(*all_arguments)
+      when 0..3
+        arguments = all_arguments.take(callback.arity)
```

Comment:
When I was looking at this, @soveran, is that one couldn’t necessarily count on it being solvable in the application. If you are injecting micromachine in a Sequel model, you have `#changed?` on the model; with an ActiveRecord model, you have a similar method. I injected micromachine in something that just does PORO and just wants micromachine to keep track of the problem.

That said, I think I was able to work around the problem…but it left me somewhat unsatisfied. Note that both #27 and this PR fix the problem I described in #26 in that the callback mechanism was broken for folks who upgraded.